### PR TITLE
convert type members

### PR DESCRIPTION
**Pull request by xeno-by**
Supersedes https://github.com/scalameta/paradise/pull/96.

### CODE REVIEW

File: tests/converter/src/test/scala/Syntactic.scala [hunk 1]
```diff
@@ -104,6 +104,18 @@ class Syntactic extends ConverterSuite {
...
 
+  // definitions
+  syntactic("type Age = Int")
+  syntactic("type Age")
+  syntactic("type Age >: Int <: Any")
+  syntactic("type Age <: Int + Boolean")
+  syntactic("type Container[T]")
+  syntactic("type Container[T] = List[T]")
+  syntactic("type Container[T] <: List[T] { def isEmpty: Boolean; type M }")
+  syntactic("type Container[T] <: List[T] with Set[T] { def isEmpty: Boolean; type M }")
```
Comment (olafurpg) [inline tests/converter/src/test/scala/Syntactic.scala]: Wow, amazing to see this work!

File: tests/converter/src/main/scala/ConverterSuite.scala [hunk 2]
```diff
@@ -100,7 +100,20 @@ trait ConverterSuite extends FunSuite {
...
-        val tree   = gen.mkTreeOrBlock(newUnitParser(code, "<toolbox>").parseStatsOrPackages())
+        val tree = {
+          // NOTE: `parseStatsOrPackages` fails to parse abstract type defs without bounds,
+          // so we need to apply a workaround to ensure that we correctly process those.
+          def somewhatBrokenParse(code: String) =
+            gen.mkTreeOrBlock(newUnitParser(code, "<toolbox>").parseStatsOrPackages())
+          val rxAbstractTypeNobounds = """^type (\w+)(\[[^=]*?\])?$""".r
+          code match {
+            case rxAbstractTypeNobounds(_ *) =>
+              val tdef @ TypeDef(mods, name, tparams, _) = somewhatBrokenParse(code + " <: Dummy")
```
Comment (olafurpg) [inline tests/converter/src/main/scala/ConverterSuite.scala]: This makes me want to cry :'( (and appreciate scala.meta even more!)

File: plugin/src/main/scala/org/scalameta/paradise/converters/LogicalTrees.scala [hunk 3]
```diff
@@ -713,7 +713,13 @@ class LogicalTrees[G <: Global](val global: G, root: G#Tree) extends ReflectTool
   object TypeDef {
     def unapply(
         tree: g.TypeDef): Option[(List[l.Modifier], l.TypeName, List[l.TypeParamDef], g.Tree)] = {
-      ???
+      tree match {
+        case g.TypeDef(mods, name, tparams, rhs) if !mods.hasFlag(DEFERRED) =>
```
Comment (olafurpg) [inline plugin/src/main/scala/org/scalameta/paradise/converters/LogicalTrees.scala]: What does the DEFERRED flag mean?

Comment (xeno-by) [inline plugin/src/main/scala/org/scalameta/paradise/converters/LogicalTrees.scala]: Abstractness.

In scalac, there are two flags with similar meaning: ABSTRACT and DEFERRED. The former is used whenever you explicitly write `abstract` (i.e. in abstract classes). The latter is used whenever a member is abstract, but you don't explicitly write `abstract` (i.e. everywhere else).

**Pull request merged**